pkg/uefi/nvram.go

```diff
@@ -0,0 +1,458 @@
+		lastVariableFlag = 0xFFFFFF
+	} else if Attributes.ErasePolarity == 0 {
+		lastVariableFlag = 0
+	} else {
+		return fmt.Errorf("erase polarity not 0x00 or 0xFF, got %#x", Attributes.ErasePolarity)
+	}
+
+	// Add next node information
+	next := Read3Size(v.Header.Next)
+	//fmt.Printf("JVDG: %02x %02x %02x  %06x\n", v.Header.Next[0], v.Header.Next[1], v.Header.Next[2], next)
+	if next != lastVariableFlag {
+		v.Type = LinkNVarEntry
+		v.NextOffset = v.Offset + next
+	}
+	return nil
+}
+
+func (v *NVar) parseExtendedHeader() error {
+	var knownExtDataFormat bool
+	if v.Header.Attributes&NVarEntryExtHeader == 0 {
+		return nil
+	}
+	var extendedHeaderSize uint16
+	r := bytes.NewReader(v.buf)
+	if _, err := r.Seek(-int64(binary.Size(extendedHeaderSize)), io.SeekEnd); err != nil {
+		return err
+	}
+	if err := binary.Read(r, binary.LittleEndian, &extendedHeaderSize); err != nil {
+		return err
+	}
+	// Sanity check extendedHeaderSize < body size
+	bodySize := int64(v.Header.Size) - v.DataOffset
+	if int64(extendedHeaderSize) > bodySize {
+		return fmt.Errorf("extended header size (0x%X) is greater than body size (0x%X)", extendedHeaderSize, bodySize)
+	}
+	v.ExtOffset = int64(v.Header.Size) - int64(extendedHeaderSize)
+	if _, err := r.Seek(v.ExtOffset, io.SeekStart); err != nil {
+		return err
+	}
+	var extAttributes NVarExtAttribute
+	if err := binary.Read(r, binary.LittleEndian, &extAttributes); err != nil {
+		return err
+	}
+	v.ExtAttributes = &extAttributes
+	// Variable with checksum
+	if extAttributes&NVarEntryExtChecksum != 0 {
+		// Get stored checksum
+		var storedChecksum uint8
+		storedChecksum = v.buf[int64(v.Header.Size)-int64(binary.Size(extendedHeaderSize)+binary.Size(storedChecksum))]
+		v.Checksum = &storedChecksum
+		// Recalculate checksum for the variable
+		calculatedChecksum := uint8(0)
+		// [4-5] Include entry size and flags
+		// [ 9 ] Include entry attributes
+		// [10-] Include entry data
+		for i := int64(4); i < int64(v.Header.Size); i++ {
+			calculatedChecksum += v.buf[i]
+			if i == 5 {
+				i += 3 // Skip Next
+			}
+		}
+		if calculatedChecksum != 0 {
+			calculatedChecksum = -calculatedChecksum
+			v.ExpectedChecksum = &calculatedChecksum
+		}
+		knownExtDataFormat = true
+	}
+
+	if v.Header.Attributes&NVarEntryAuthWrite == 0 {
+		var timestamp uint64
+		if err := binary.Read(r, binary.LittleEndian, &timestamp); err != nil {
+			switch err {
+			case io.EOF, io.ErrUnexpectedEOF:
+				return fmt.Errorf("extended header size (0x%X) is too small for timestamp", extendedHeaderSize)
+			default:
+				return err
+			}
+		}
+		if v.Header.Attributes&NVarEntryDataOnly != 0 {
+			// Full or link variable have hash
+			hashstart, err := r.Seek(0, io.SeekCurrent)
+			if err != nil {
+				return err
+			}
+			if hashstart+sha256.Size > int64(v.Header.Size) {
+				return fmt.Errorf("extended header size (0x%X) is too small for hash", extendedHeaderSize)
+			}
+			v.Hash = v.buf[hashstart : hashstart+sha256.Size]
```

Comment:
Suggest that you do a copy here in case the buffer is ever modified.

Reply:
done